### PR TITLE
Feat/talk2scholars zotero read

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

https://github.com/user-attachments/assets/2286efea-03c0-453f-9696-8d6d55795d88

## Description - Initial Integration of Zotero Agent
This PR introduces the **Zotero Agent**, marking its initial integration into the system. At this stage, the agent includes a single tool:  
- **`zotero_read`**: Enables searching for papers within the Zotero library of the user.  

- The fetched results can later be queried using the existing tools of `s2_agent`, showcasing true integration capabilities. The `query_results` tool, which contains the `DataFrame agent`, is used to query the last displayed article, and the `display_results` tool from the `s2_agent` universe is used to display the artifact. Additionally, we have the capability to search for and recommend papers based on the retrieved results from the Zotero library. Here, we use our `retrieve_semantic_scholar_id` tool from the previous PR #106, which can now fetch the Semantic Scholar ID of the title that the user selects from the last displayed paper. This ID is then passed to our search and recommendation tools.

- Right now we have an argument of `query` which is a really cool feature, through we can directly put a search term and get results for it.
Example: "Can you find me paper on biology from my zotero library" - as shown in the demo

#### **Filtering capabilities**:  
  - Currently, we filter results based on item type, later we plan to use `Literal[*options]`, so that user can give in a specific term:  
    ```python
    filter_item_types = (
        ["journalArticle", "conferencePaper", "preprint"] if only_articles else []
    )
    ```
  - Additional filters will be introduced in subsequent updates.

- **Zotero Library Key Handling**:  
  - Currently, Zotero credentials (Library Key & User ID) are set via environment variables, similar to OpenAI API keys.  
  - A future update will add **Streamlit UI integration**, allowing users to log in directly and use the application more easily.  

#### **Future Plans**
This PR serves as the foundation for further enhancements to the Zotero Agent. The next steps include:
- Introducing `zotero_write`: Allowing user to write the searched/recommended paper within the Zotero library.  
- Currently, The agent does not determine the path of the paper, the agent considers all the path and shows all the paper, so we have to add a columns to display path

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Comment by ansh-info** [comment]
@dmccloskey 

I have a general question about **Code Quality & Testing**

Currently, all tests are passing with full code coverage, Right now Im transitioning to **tests with real LLM evaluation(using openai)** instead of relying solely on **mocked tests**. These new tests will serve as **integration tests**, providing a more accurate assessment of the application’s behavior and performance in real-world scenarios. This shift will ensure more **robust and reliable testing** for the system.
So is it the right choice to completely move to openai/real llm tests, or keep the mock tests as well? or a 50-50 Split

**Comment by dmccloskey** [comment]
@ansh-info Good question. The choice between Mock vs. integration testing with an LLM comes down to what is being tested. 

If we are speaking about application logic which is independent of the LLM, (or other external APIs/services), then Mocking is sufficient. In our particular case, the application logic should be independent of the LLM and external services so Mocking should be sufficient for the majority of testing that we do.

However, once we are concerned about how the application behaves based on particular prompts or input data from our users, then we do need to incorporate the LLM and external services to ensure our application logic behaves as expected. For example, if we are interested in tuning prompts to improve the response of the application to different user queries, then we need to incorporate an LLM. In another example, we would want to ensure that our application can fulfil our intended user journeys, which would require as realistic data as possible and then an LLM to complete the application. The latter is also important when it comes to testing different LLMs so that we have a baseline for the expected response based on defined prompts and user input.

My recommendation would be the following: ensure the application logic is as independent as possible of the LLM so that we can use Mocking (which is also cheaper. faster, and less burdensome on other developers) and we save the integration tests with LLMs for prompt tuning (or other application features that are dependent upon the LLM) and end-to-end user journeys.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The ability to retrieve papers from Zotero looks really cool 😎. I had a few comments related to the tests, prompts, and configs. Also, I notice the format for the year on the demo video was formatted as number "2,022" and not a date "2022".

File: aiagents4pharma/talk2scholars/tests/test_call_s2.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    # Use the real OpenAI LLM
+    llm = ChatOpenAI(model="gpt-4o-mini", temperature=hydra_cfg.temperature)
+
+    # Initialize main agent workflow with real LLM
+    thread_id = "test_thread"
+    app = get_app(thread_id, llm)
+
+    # Provide an actual user query
+    initial_state = Talk2Scholars(
+        messages=[HumanMessage(content="Find AI papers on reinforcement learning")]
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_call_s2.py]: I've incorporated synonym variations into the test suite to ensure that different phrasings (e.g., 'find' → 'search', 'locate') are correctly mapped to the appropriate agent.

File: aiagents4pharma/talk2scholars/tests/test_call_s2.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from langchain_openai import ChatOpenAI
+from langchain_core.messages import HumanMessage, AIMessage
+from ..agents.main_agent import get_app
+from ..state.state_talk2scholars import Talk2Scholars
+
+
+@pytest.mark.skipif(
+    not os.getenv("OPENAI_API_KEY"), reason="Requires OpenAI API key to run"
+)
+def test_call_s2_agent_with_real_llm():
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_call_s2.py]: I've updated the routing logic tests to remove the dependency on an LLM. Instead, the tests now use direct query matching with a mock supervisor node to ensure deterministic behaviour.

File: aiagents4pharma/talk2scholars/configs/agents/talk2scholars/s2_agent/default.yaml [hunk 3]
```diff
@@ -8,9 +8,121 @@ temperature: 0
 s2_agent: >
   You are an academic research assistant with access to the
   Semantic Scholar API for paper discovery and analysis.
+
+  AVAILABLE TOOLS:
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/agents/talk2scholars/s2_agent/default.yaml:12]: **Thank you for your review and feedback!**  

I have removed the tool arguments and made the prompt significantly smaller. Initially, these arguments were included to reduce LLM hallucinations, but I had planned to remove them in the future for a more robust solution.  

Additionally, I have implemented a new feature (proposed by @gurdeep330 ) where I pass the **top 3 results** from the API response as content to the LLM. This provides relevant context and helps mitigate hallucination. Along with the **title**, I now include the **year** for the Semantic Scholar tools. For the Zotero tool, I also provide the **filter type** as additional context.  

Let me know if you have any further suggestions!

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_read.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        query,
+        only_articles,
+        limit,
+    )
+
+    # Initialize Zotero client
+    zot = zotero.Zotero(cfg.user_id, cfg.library_type, cfg.api_key)
+
+    # Get items matching the query
+    items = zot.items(q=query, limit=min(limit, 100))
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_read.py]: Sure, I moved it to Hydra

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_read.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    logger.info("Received %d items from Zotero", len(items))
+
+    # Define filter criteria
+    filter_item_types = (
+        ["journalArticle", "conferencePaper", "preprint"] if only_articles else []
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_read.py]: It was originally planned to be in Hydra, but I had temporarily commented it out. I have now moved it back to Hydra for better configuration management.

**Comment by ansh-info** [comment]
> The ability to retrieve papers from Zotero looks really cool 😎. I had a few comments related to the tests, prompts, and configs. Also, I notice the format for the year on the demo video was formatted as number "2,022" and not a date "2022".

Thank you for the feedback
I have fixed it, please have a look 
image

**Comment by ansh-info** [comment]
> @ansh-info Good question. The choice between Mock vs. integration testing with an LLM comes down to what is being tested.
> 
> If we are speaking about application logic which is independent of the LLM, (or other external APIs/services), then Mocking is sufficient. In our particular case, the application logic should be independent of the LLM and external services so Mocking should be sufficient for the majority of testing that we do.
> 
> However, once we are concerned about how the application behaves based on particular prompts or input data from our users, then we do need to incorporate the LLM and external services to ensure our application logic behaves as expected. For example, if we are interested in tuning prompts to improve the response of the application to different user queries, then we need to incorporate an LLM. In another example, we would want to ensure that our application can fulfil our intended user journeys, which would require as realistic data as possible and then an LLM to complete the application. The latter is also important when it comes to testing different LLMs so that we have a baseline for the expected response based on defined prompts and user input.
> 
> My recommendation would be the following: ensure the application logic is as independent as possible of the LLM so that we can use Mocking (which is also cheaper. faster, and less burdensome on other developers) and we save the integration tests with LLMs for prompt tuning (or other application features that are dependent upon the LLM) and end-to-end user journeys.

@dmccloskey  Thank you for the detailed explanation! That makes a lot of sense. I now understand the importance of keeping the application logic independent of the LLM where possible and using mocking for most tests. I'll ensure that integration tests with real LLMs are used primarily for prompt tuning and end-to-end user journeys. Appreciate the guidance!

File: aiagents4pharma/talk2scholars/tests/test_call_s2.py [hunk 6]
```diff
@@ -1,56 +1,104 @@
...
+            ),  # Ensure user query is retained
+            AIMessage(
+                content="Found relevant papers on deep learning."
+            ),  # Ensure AI response is added
+        ],
+        "papers": {"paper1": "Paper on deep learning"},
+        "multi_papers": {},
+        "last_displayed_papers": {},
+    }
+    print(
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_call_s2.py:41]: I would recommend removing these print statements if they are no longer needed.

File: aiagents4pharma/talk2scholars/tests/test_call_zotero.py [hunk 7]
```diff
@@ -1,56 +1,58 @@
...
+    assert "Retrieve my Zotero papers." in result_messages  # User query
+    assert any("Zotero" in msg for msg in result_messages)  # Zotero response expected
+    assert "zotero_read" in result and isinstance(
+        result["zotero_read"], dict
+    )  # Data exists
 
-    # Assert that the supervisor routed correctly
-    assert "messages" in result, "Expected messages in response"
+    # Ensure logs confirm `call_zotero_agent` was invoked
+    assert "Calling Zotero agent" in caplog.text  # Log entry before calling
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_call_zotero.py:55]: I do not recommend using the logs in testing. I would think about another way to verify this.

File: aiagents4pharma/talk2scholars/tests/test_routing_logic.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@pytest.fixture
+def mock_state():
+    """Creates a mock state to simulate an ongoing conversation."""
+    return Talk2Scholars(messages=[])
+
+
+@pytest.fixture
+def mock_router():
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_routing_logic.py:21]: Nice 👍

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Getting close 👍. I had a few comments on the tests.

File: aiagents4pharma/talk2scholars/tests/test_call_zotero.py (hunk 7 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_call_zotero.py:55]: I've Updated the tests and stopped using logs for testing and now check if the Zotero response and data are correctly added to the conversation

**Review by dmccloskey — APPROVED** [review]
I can't wait to give the zotero feature a test drive 😀. Nice work 💪

**Comment by dmccloskey** [comment]
Hi @ansh-info It looks like there is an issue with the requirements that are causing the tests to fail. Can you take a look and see what the issue is?

File: requirements.txt
Comment (gurdeep330) [inline requirements.txt:1]: @ansh-info, could you please undo these changes?
I'd recommend not to use pip freeze > requirements.txt
This causes the installation of unwanted packages. Please enter the new package name manually to the requirements.txt file.

In case there are compatibility issues of pyzotero with Ollama, please open a new discussion and tag Ahmad. Once he's OK with the new version of Ollama, you can come back and update the requirements.txt in this PR.

**Review by gurdeep330 — CHANGES_REQUESTED** [review]

**Comment by ansh-info** [comment]
> Hi @ansh-info It looks like there is an issue with the requirements that are causing the tests to fail. Can you take a look and see what the issue is?

This issue was caused by a dependency conflict between ollama and httpx. ollama 0.4.6 requires httpx<0.28.0, but pyzotero needs httpx>=0.28.1, which led to an incompatibility during installation.
I didn’t know that you don’t have to do pip freeze > requirements.txt. I will take care of this next time. I’m sorry!

File: requirements.txt
Comment (ansh-info) [reply requirements.txt:1]: Got it! I’ve undone the changes and will manually add new packages to requirements.txt moving forward. I now understand that pip freeze > requirements.txt can add unnecessary dependencies.
Regarding the compatibility issue between pyzotero and ollama, I’ll open a discussion and tag Ahmad. Once we get approval on the new ollama version, I’ll update requirements.txt accordingly in this PR.
Thanks for the guidance!

**Comment by gurdeep330** [comment]
> > Hi @ansh-info It looks like there is an issue with the requirements that are causing the tests to fail. Can you take a look and see what the issue is?
> 
> This issue was caused by a dependency conflict between ollama and httpx. ollama 0.4.6 requires httpx<0.28.0, but pyzotero needs httpx>=0.28.1, which led to an incompatibility during installation.
> I didn’t know that you don’t have to do pip freeze > requirements.txt. I will take care of this next time. I’m sorry!

No worries. Hopefully, we can get it sorted by tomorrow. Thx for your patience 🙏

**Review by gurdeep330 — APPROVED** [review]
All the pylint workflows passed. Thanks for fixing this @ansh-info !
